crates/order-validation/src/banned.rs

```diff
@@ -1,22 +1,128 @@
 use {
-    cached::{Cached, TimedCache},
     contracts::ChainalysisOracle,
     ethcontract::{errors::MethodError, futures::future::join_all, H160},
-    std::{collections::HashSet, sync::Mutex},
+    std::{
+        collections::{HashMap, HashSet},
+        ops::Div,
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+    tokio::sync::RwLock,
 };
 
 /// A list of banned users and an optional registry that can be checked onchain.
 pub struct Users {
     list: HashSet<H160>,
-    onchain: Option<Onchain>,
+    onchain: Option<Arc<Onchain>>,
+}
+
+#[derive(Clone)]
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+}
+
+impl UserMetadata {
+    pub fn with_banned(mut self, banned: bool) -> Self {
+        self.is_banned = banned;
+        self
+    }
+
+    pub fn with_last_updated(mut self, last_updated: Instant) -> Self {
+        self.last_updated = last_updated;
+        self
+    }
 }
 
 struct Onchain {
     contract: ChainalysisOracle,
-    cache: Mutex<TimedCache<H160, bool>>,
+    cache: RwLock<HashMap<H160, UserMetadata>>,
 }
 
-const TTL: u64 = 60 * 60; // 1 hour
+impl Onchain {
+    pub fn new(contract: ChainalysisOracle) -> Arc<Self> {
+        let onchain = Arc::new(Self {
+            contract,
+            cache: Default::default(),
+        });
+
+        onchain.clone().spawn_maintenance_task();
+
+        onchain
+    }
+
+    /// Spawns a background task that periodically checks the cache for expired
+    /// entries and re-run checks for them.
+    ///
+    /// Removes from the cache entries that are expired and not limit order
+    /// participants and not part of the current auction. This is made to keep
```

Comment:
I don't see where in the code we check if this is a limit order (I also don't think we should, but instead keep all users up to date, it's not that many records)